#             best_acc = soul.evolve(X_evo_v, y_evo_v, generations=50)


Increase gen for Stability and accuracy.

In [72]:
import numpy as np
import pandas as pd
import warnings
import random
from scipy.optimize import minimize
from scipy.fft import fft

# Sklearn Core & Metrics
from sklearn.base import BaseEstimator, ClassifierMixin
from sklearn.utils.validation import check_X_y, check_array, check_is_fitted
from sklearn.metrics import log_loss, accuracy_score
from sklearn.model_selection import StratifiedKFold, train_test_split
from sklearn.preprocessing import RobustScaler, PowerTransformer, StandardScaler

# Sklearn Transformers
from sklearn.random_projection import GaussianRandomProjection
from sklearn.decomposition import PCA
from sklearn.kernel_approximation import RBFSampler

# Sklearn Models
from sklearn.ensemble import ExtraTreesClassifier, RandomForestClassifier, HistGradientBoostingClassifier
from sklearn.linear_model import RidgeClassifier, Ridge
from sklearn.mixture import GaussianMixture
from sklearn.svm import SVC, NuSVC, LinearSVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis, LinearDiscriminantAnalysis
from sklearn.kernel_ridge import KernelRidge
from sklearn.calibration import CalibratedClassifierCV

# Gradient Boosting
from xgboost import XGBClassifier

# GPU CHECK
try:
    import cupy as cp
    GPU_AVAILABLE = True
    print("✅ GPU DETECTED: HRF v26.0 'Holo-Fractal Universe' Active")
except ImportError:
    GPU_AVAILABLE = False
    print("⚠️ GPU NOT FOUND: Running in Slow Mode")

warnings.filterwarnings('ignore')

# --- 1. THE HOLOGRAPHIC SOUL (Unit 3 - Multiverse Edition) ---
class HolographicSoulUnit(BaseEstimator, ClassifierMixin):
    def __init__(self, k=15):
        self.k = k
        self.dna_ = {
            'freq': 2.0, 'gamma': 0.5, 'power': 2.0,
            'metric': 'minkowski', 'p': 2.0,
            'phase': 0.0, 'dim_reduction': 'none'
        }
        self.projector_ = None
        self.X_raw_source_ = None

    def fit(self, X, y):
        self.classes_ = np.unique(y)
        self._apply_projection(X)
        self.y_train_ = y
        return self

    def _apply_projection(self, X):
        if self.dna_['dim_reduction'] == 'holo':
            n_components = max(2, int(np.sqrt(X.shape[1])))
            self.projector_ = GaussianRandomProjection(n_components=n_components, random_state=42)
            self.X_train_ = self.projector_.fit_transform(X)
        elif self.dna_['dim_reduction'] == 'pca':
            n_components = max(2, int(np.sqrt(X.shape[1])))
            self.projector_ = PCA(n_components=n_components, random_state=42)
            self.X_train_ = self.projector_.fit_transform(X)
        else:
            self.projector_ = None
            self.X_train_ = X

    def set_raw_source(self, X):
        self.X_raw_source_ = X

    def evolve(self, X_val, y_val, generations=1000):
        n_universes = 10
        best_acc = self.score(X_val, y_val)
        best_dna = self.dna_.copy()

        # Smart Init
        if GPU_AVAILABLE:
            sample_X = cp.asarray(self.X_train_[:100])
            dists = cp.mean(cp.linalg.norm(sample_X[:, None, :] - sample_X[None, :, :], axis=2))
            median_dist = float(cp.asnumpy(dists))
        else:
            median_dist = 1.0

        if median_dist > 0:
            best_dna['freq'] = 3.14159 / median_dist

        for i in range(generations):
            candidates = []
            for _ in range(n_universes):
                mutant = best_dna.copy()
                trait = random.choice(list(mutant.keys()))
                if trait == 'freq': mutant['freq'] *= np.random.uniform(0.8, 1.25)
                elif trait == 'gamma': mutant['gamma'] = np.random.uniform(0.1, 5.0)
                elif trait == 'power': mutant['power'] = random.choice([0.5, 1.0, 2.0, 3.0, 4.0, 6.0])
                elif trait == 'p': mutant['p'] = np.clip(mutant['p'] + np.random.uniform(-0.5, 0.5), 0.5, 8.0)
                elif trait == 'phase': mutant['phase'] = np.random.uniform(0, 3.14159)
                elif trait == 'dim_reduction': mutant['dim_reduction'] = random.choice(['none', 'holo', 'pca'])
                candidates.append(mutant)

            generation_best_acc = -1
            generation_best_dna = None

            for mutant_dna in candidates:
                self.dna_ = mutant_dna
                if self.X_raw_source_ is not None: self._apply_projection(self.X_raw_source_)
                acc = self.score(X_val, y_val)
                if acc > generation_best_acc:
                    generation_best_acc = acc
                    generation_best_dna = mutant_dna

            if generation_best_acc >= best_acc:
                best_acc = generation_best_acc
                best_dna = generation_best_dna
            else:
                self.dna_ = best_dna
                if self.X_raw_source_ is not None: self._apply_projection(self.X_raw_source_)

        self.dna_ = best_dna
        return best_acc

    def predict_proba(self, X):
        if self.projector_ is not None: X_curr = self.projector_.transform(X)
        else: X_curr = X
        if GPU_AVAILABLE: return self._predict_proba_gpu(X_curr)
        else: return np.zeros((len(X), len(self.classes_)))

    def _predict_proba_gpu(self, X):
        X_tr_g = cp.asarray(self.X_train_, dtype=cp.float32)
        X_te_g = cp.asarray(X, dtype=cp.float32)
        y_tr_g = cp.asarray(self.y_train_)

        n_test = len(X_te_g)
        n_classes = len(self.classes_)
        probas = []
        batch_size = 256

        p_norm = self.dna_.get('p', 2.0)
        gamma = self.dna_['gamma']
        freq = self.dna_['freq']
        power = self.dna_['power']
        phase = self.dna_.get('phase', 0.0)

        for i in range(0, n_test, batch_size):
            end = min(i + batch_size, n_test)
            batch_te = X_te_g[i:end]
            diff = cp.abs(batch_te[:, None, :] - X_tr_g[None, :, :])
            dists = cp.sum(cp.power(diff, p_norm), axis=2)
            dists = cp.power(dists, 1.0/p_norm)
            top_k_idx = cp.argsort(dists, axis=1)[:, :self.k]
            row_idx = cp.arange(len(batch_te))[:, None]
            top_dists = dists[row_idx, top_k_idx]
            top_y = y_tr_g[top_k_idx]

            cosine_term = 1.0 + cp.cos(freq * top_dists + phase)
            cosine_term = cp.maximum(cosine_term, 0.0)
            w = cp.exp(-gamma * (top_dists**2)) * cosine_term
            w = cp.power(w, power)

            batch_probs = cp.zeros((len(batch_te), n_classes))
            for c_idx, cls in enumerate(self.classes_):
                class_mask = (top_y == cls)
                batch_probs[:, c_idx] = cp.sum(w * class_mask, axis=1)

            total_energy = cp.sum(batch_probs, axis=1, keepdims=True)
            total_energy[total_energy == 0] = 1.0
            batch_probs /= total_energy
            probas.append(batch_probs)
            del batch_te, dists, diff, top_k_idx, top_dists, w, cosine_term
            cp.get_default_memory_pool().free_all_blocks()

        return cp.asnumpy(cp.concatenate(probas))

    def predict(self, X):
        return self.classes_[np.argmax(self.predict_proba(X), axis=1)]

    def score(self, X, y):
        return accuracy_score(y, self.predict(X))

# --- 3. THE QUANTUM FIELD (Unit 4 - Reserve) ---
class QuantumFieldUnit(BaseEstimator, ClassifierMixin):
    def __init__(self):
        self.rbf_feature_ = RBFSampler(n_components=100, random_state=42)
        self.classifier_ = RidgeClassifier(alpha=1.0)
        self.classes_ = None
        self.dna_ = {'gamma': 1.0, 'n_components': 100}

    def fit(self, X, y):
        self.classes_ = np.unique(y)
        self.rbf_feature_.set_params(gamma=self.dna_['gamma'], n_components=self.dna_['n_components'])
        X_quantum = self.rbf_feature_.fit_transform(X)
        self.classifier_.fit(X_quantum, y)
        return self

    def predict_proba(self, X):
        X_quantum = self.rbf_feature_.transform(X)
        d = self.classifier_.decision_function(X_quantum)
        if len(self.classes_) == 2:
            probs = 1 / (1 + np.exp(-d))
            return np.column_stack([1-probs, probs])
        else:
            exp_d = np.exp(d - np.max(d, axis=1, keepdims=True))
            return exp_d / np.sum(exp_d, axis=1, keepdims=True)

    def score(self, X, y):
        return accuracy_score(y, self.classes_[np.argmax(self.predict_proba(X), axis=1)])

# --- 4. THE ENTROPY MAXWELL (Unit 5 - Reserve) ---
class EntropyMaxwellUnit(BaseEstimator, ClassifierMixin):
    def __init__(self):
        self.models_ = {}
        self.classes_ = None
        self.priors_ = None
        self.dna_ = {'n_components': 1}

    def fit(self, X, y):
        self.classes_ = np.unique(y)
        self.models_ = {}
        self.priors_ = {}
        n_samples = len(y)
        for cls in self.classes_:
            X_c = X[y == cls]
            if len(X_c) < 2:
                self.priors_[cls] = 0.0
                continue
            self.priors_[cls] = len(X_c) / n_samples
            n_comp = min(self.dna_['n_components'], len(X_c))
            gmm = GaussianMixture(n_components=n_comp, covariance_type='full',
                                  reg_covar=1e-4, random_state=42)
            gmm.fit(X_c)
            self.models_[cls] = gmm
        return self

    def predict_proba(self, X):
        probs = np.zeros((len(X), len(self.classes_)))
        for i, cls in enumerate(self.classes_):
            if cls in self.models_:
                log_prob = self.models_[cls].score_samples(X)
                log_prob = np.clip(log_prob, -100, 100)
                probs[:, i] = np.exp(log_prob) * self.priors_[cls]
        total = np.sum(probs, axis=1, keepdims=True) + 1e-10
        return probs / total

    def score(self, X, y):
        return accuracy_score(y, self.classes_[np.argmax(self.predict_proba(X), axis=1)])

# --- 5. THE OMNI-KERNEL NEXUS (Unit 6 - Reserve) ---
class OmniKernelUnit(BaseEstimator, ClassifierMixin):
    def __init__(self):
        self.model_ = None
        self.classes_ = None
        self.dna_ = {'kernel': 'rbf', 'C': 1.0, 'gamma': 'scale', 'degree': 3, 'coef0': 0.0}

    def fit(self, X, y):
        self.classes_ = np.unique(y)
        self.model_ = SVC(
            kernel=self.dna_['kernel'], C=self.dna_['C'], gamma=self.dna_['gamma'],
            degree=self.dna_['degree'], coef0=self.dna_['coef0'],
            probability=True, random_state=42, cache_size=500
        )
        self.model_.fit(X, y)
        return self

    def predict_proba(self, X):
        return self.model_.predict_proba(X)

    def score(self, X, y):
        return self.model_.score(X, y)

# --- 18. THE GOLDEN SPIRAL (Unit 18 - Nature's Code) ---
class GoldenSpiralUnit(BaseEstimator, ClassifierMixin):
    def __init__(self, k=21): # k=21 is a Fibonacci number
        self.k = k
        self.PHI = 1.6180339887
        self.classes_ = None
        self.X_train_ = None
        self.y_train_ = None
        # DNA: The mutable parameters of the spiral
        self.dna_ = {
            'resonance': 1.618,  # The spiral tightness
            'decay': 1.0,        # How fast influence drops
            'shift': 0.0         # Rotation of the spiral
        }

    def fit(self, X, y):
        self.classes_ = np.unique(y)
        self.X_train_ = np.array(X, dtype=np.float32)
        self.y_train_ = np.array(y)
        return self

    def evolve(self, X_val, y_val, generations=50):
        """
        Self-evolves the spiral shape to maximize harmony (accuracy)
        on the validation set.
        """
        best_acc = self.score(X_val, y_val)
        best_dna = self.dna_.copy()

        # Try mutating the spiral geometry
        for _ in range(generations):
            mutant = best_dna.copy()
            trait = random.choice(['resonance', 'decay', 'shift'])

            if trait == 'resonance':
                # Mutate around Phi
                mutant['resonance'] *= np.random.uniform(0.9, 1.1)
            elif trait == 'decay':
                mutant['decay'] = np.random.uniform(0.5, 3.0)
            elif trait == 'shift':
                mutant['shift'] += np.random.uniform(-0.5, 0.5)

            self.dna_ = mutant
            acc = self.score(X_val, y_val)

            if acc > best_acc:
                best_acc = acc
                best_dna = mutant
            else:
                self.dna_ = best_dna # Revert

        return best_acc

    def predict_proba(self, X):
        # We implement the Phi-Weighted KNN manually for speed/control
        if GPU_AVAILABLE:
            return self._predict_proba_gpu(X)

        n_test = len(X)
        n_classes = len(self.classes_)
        probs = np.zeros((n_test, n_classes))

        # CPU Slow Path (Fallback)
        # Note: For production, use GPU path or optimized cdist
        for i, x_query in enumerate(X):
            # 1. Phi-Minkowski Distance (p = 1.618)
            diff = np.abs(self.X_train_ - x_query)
            dists = np.sum(diff ** self.PHI, axis=1) ** (1/self.PHI)

            # 2. Find Fibonacci Neighbors
            idx = np.argsort(dists)[:self.k]
            nearest_dists = dists[idx]
            nearest_y = self.y_train_[idx]

            # 3. The Golden Spiral Kernel
            # Weight = 1 / (d^phi) * Cosine_Resonance
            # This mimics constructive interference along a spiral
            w = (1.0 / (nearest_dists ** self.PHI + 1e-9))

            # Add Logarithmic Spiral Rotation
            # Nature grows logarithmically: r = a * e^(k * theta)
            w *= (1.0 + 0.5 * np.cos(np.log(nearest_dists + 1e-9) * self.dna_['resonance'] + self.dna_['shift']))
            w = np.maximum(w, 0) # Remove destructive interference (negative weights)

            for c_i, cls in enumerate(self.classes_):
                probs[i, c_i] = np.sum(w[nearest_y == cls])

        # Normalize
        sums = np.sum(probs, axis=1, keepdims=True)
        sums[sums == 0] = 1
        return probs / sums

    def _predict_proba_gpu(self, X):
        import cupy as cp
        X_tr_g = cp.asarray(self.X_train_)
        X_te_g = cp.asarray(X)
        y_tr_g = cp.asarray(self.y_train_)

        n_test = len(X)
        n_classes = len(self.classes_)
        probas = []
        batch_size = 256 # Process in chunks

        phi = self.PHI
        res = self.dna_['resonance']
        shift = self.dna_['shift']

        for i in range(0, n_test, batch_size):
            end = min(i+batch_size, n_test)
            batch = X_te_g[i:end]

            # Phi-Minkowski Distance
            diff = cp.abs(batch[:, None, :] - X_tr_g[None, :, :])
            dists = cp.sum(cp.power(diff, phi), axis=2)
            dists = cp.power(dists, 1.0/phi)

            # Get Top K
            top_k_idx = cp.argsort(dists, axis=1)[:, :self.k]
            row_idx = cp.arange(len(batch))[:, None]
            top_dists = dists[row_idx, top_k_idx]
            top_y = y_tr_g[top_k_idx]

            # Golden Spiral Kernel Calculation
            # 1. Power Law Decay based on Phi
            base_w = 1.0 / (cp.power(top_dists, phi) + 1e-9)

            # 2. Logarithmic Spiral Modulation
            # This creates "shells" of probability rather than a flat circle
            spiral_mod = 1.0 + 0.5 * cp.cos(cp.log(top_dists + 1e-9) * res + shift)

            w = base_w * cp.maximum(spiral_mod, 0.0)

            batch_p = cp.zeros((len(batch), n_classes))
            for c_idx, cls in enumerate(self.classes_):
                mask = (top_y == cls)
                batch_p[:, c_idx] = cp.sum(w * mask, axis=1)

            probas.append(batch_p)

            # Clean up GPU memory
            del batch, diff, dists, top_k_idx, top_dists, w
            cp.get_default_memory_pool().free_all_blocks()

        final_p = cp.asnumpy(cp.concatenate(probas))

        # Normalize
        sums = np.sum(final_p, axis=1, keepdims=True)
        sums[sums == 0] = 1.0
        return final_p / sums

    def score(self, X, y):
        return accuracy_score(y, self.classes_[np.argmax(self.predict_proba(X), axis=1)])



# --- 19. THE ENTROPY MAXWELL (Thermodynamics) ---
# --- 19. THE ENTROPY MAXWELL (Thermodynamics - Evolving) ---
class EntropyMaxwellUnit(BaseEstimator, ClassifierMixin):
    def __init__(self, n_components=1):
        self.n_components = n_components
        self.dna_ = {'n_components': n_components} # DNA storage
        self.models_ = {}
        self.priors_ = {}
        self.classes_ = None

    def fit(self, X, y):
        self.classes_ = np.unique(y)
        # Use DNA parameter instead of static init parameter
        active_n = self.dna_['n_components']

        for cls in self.classes_:
            X_c = X[y == cls]
            # Safety: Cannot have more components than samples
            nc = min(active_n, len(X_c)) if len(X_c) > 1 else 1

            if len(X_c) < 2:
                self.models_[cls] = None
                self.priors_[cls] = 0
                continue

            # Reg_covar added for stability during evolution
            gmm = GaussianMixture(n_components=nc, covariance_type='full', reg_covar=1e-3, random_state=42)
            gmm.fit(X_c)
            self.models_[cls] = gmm
            self.priors_[cls] = len(X_c) / len(y)
        return self

    def evolve(self, X_val, y_val, generations=10):
        # Thermodynamic Annealing: Try different component counts
        best_acc = self.score(X_val, y_val)
        best_dna = self.dna_.copy()

        # Try finding the optimal complexity (1 to 5 components)
        possible_components = [1, 2, 3, 4, 5]

        for comp in possible_components:
            self.dna_['n_components'] = comp
            try:
                self.fit(X_val, y_val) # Re-fit to test adaptation
                acc = self.score(X_val, y_val)
                if acc > best_acc:
                    best_acc = acc
                    best_dna = self.dna_.copy()
            except:
                continue # Skip if unstable

        self.dna_ = best_dna
        return best_acc

    def predict_proba(self, X):
        probs = np.zeros((len(X), len(self.classes_)))
        for i, cls in enumerate(self.classes_):
            if self.models_.get(cls) is not None:
                log_prob = self.models_[cls].score_samples(X)
                probs[:, i] = np.exp(log_prob) * self.priors_[cls]
        total = np.sum(probs, axis=1, keepdims=True) + 1e-10
        return probs / total

    def score(self, X, y):
        return accuracy_score(y, self.classes_[np.argmax(self.predict_proba(X), axis=1)])



# --- 20. THE QUANTUM FLUX (Superposition) ---
# --- 20. THE QUANTUM FLUX (Superposition - Evolving) ---
class QuantumFluxUnit(BaseEstimator, ClassifierMixin):
    def __init__(self, gamma=1.0, n_components=500):
        self.dna_ = {'gamma': gamma, 'n_components': n_components}
        self.rbf_feature_ = None
        self.classifier_ = RidgeClassifier(alpha=1.0)
        self.classes_ = None

    def fit(self, X, y):
        self.classes_ = np.unique(y)
        # Apply DNA settings
        self.rbf_feature_ = RBFSampler(
            gamma=self.dna_['gamma'],
            n_components=int(self.dna_['n_components']),
            random_state=42
        )
        X_quantum = self.rbf_feature_.fit_transform(X)
        self.classifier_.fit(X_quantum, y)
        return self

    def evolve(self, X_val, y_val, generations=20):
        best_acc = self.score(X_val, y_val)
        best_dna = self.dna_.copy()

        for _ in range(generations):
            mutant = best_dna.copy()
            trait = random.choice(['gamma', 'n_components'])

            if trait == 'gamma':
                mutant['gamma'] *= np.random.uniform(0.5, 2.0)
            elif trait == 'n_components':
                mutant['n_components'] = int(mutant['n_components'] * np.random.uniform(0.8, 1.2))
                mutant['n_components'] = max(50, mutant['n_components']) # Minimum floor

            self.dna_ = mutant
            self.fit(X_val, y_val) # Quick Re-fit
            acc = self.score(X_val, y_val)

            if acc > best_acc:
                best_acc = acc
                best_dna = mutant
            else:
                self.dna_ = best_dna # Revert

        return best_acc

    def predict_proba(self, X):
        X_quantum = self.rbf_feature_.transform(X)
        d = self.classifier_.decision_function(X_quantum)
        if len(self.classes_) == 2:
            prob = 1 / (1 + np.exp(-d))
            return np.column_stack([1-prob, prob])
        else:
            exp_d = np.exp(d - np.max(d, axis=1, keepdims=True))
            return exp_d / np.sum(exp_d, axis=1, keepdims=True)

    def score(self, X, y):
        return accuracy_score(y, self.classes_[np.argmax(self.predict_proba(X), axis=1)])

# --- 21. THE EVENT HORIZON (General Relativity) ---
# --- 21. THE EVENT HORIZON (General Relativity - Evolving) ---
class EventHorizonUnit(BaseEstimator, ClassifierMixin):
    def __init__(self):
        self.centroids_ = None
        self.masses_ = None
        self.schwarzschild_ = None
        self.classes_ = None
        # DNA: Laws of Physics (Gravity & Singularity limits)
        self.dna_ = {'horizon_pct': 10.0, 'decay_power': 2.0}

    def fit(self, X, y):
        self.classes_ = np.unique(y)
        self.centroids_ = []
        self.masses_ = []
        self.schwarzschild_ = []

        # Physics Parameters from DNA
        h_pct = np.clip(self.dna_['horizon_pct'], 1.0, 50.0) # Limit horizon size

        for cls in self.classes_:
            X_c = X[y == cls]
            if len(X_c) == 0: continue

            centroid = np.mean(X_c, axis=0)
            mass = len(X_c)

            dists = np.linalg.norm(X_c - centroid, axis=1)
            # Dynamic Event Horizon based on evolved percentile
            radius = np.percentile(dists, h_pct) if len(dists) > 0 else 0.0

            self.centroids_.append(centroid)
            self.masses_.append(mass)
            self.schwarzschild_.append(radius)
        return self

    def evolve(self, X_val, y_val, generations=20):
        # Evolving the fabric of spacetime
        best_acc = self.score(X_val, y_val)
        best_dna = self.dna_.copy()

        for _ in range(generations):
            mutant = best_dna.copy()
            trait = random.choice(['horizon_pct', 'decay_power'])

            if trait == 'horizon_pct':
                # Mutate the size of the black hole core
                mutant['horizon_pct'] += np.random.uniform(-5.0, 5.0)
            elif trait == 'decay_power':
                # Mutate how far gravity reaches (Newtonian = 2.0)
                mutant['decay_power'] *= np.random.uniform(0.8, 1.25)

            self.dna_ = mutant
            self.fit(X_val, y_val) # Re-calculate horizons
            acc = self.score(X_val, y_val)

            if acc > best_acc:
                best_acc = acc
                best_dna = mutant
            else:
                self.dna_ = best_dna # Revert physics

        return best_acc

    def predict_proba(self, X):
        probs = []
        decay = self.dna_['decay_power']

        for x_pt in X:
            forces = []
            in_horizon = False
            horizon_class = -1

            for i, cls in enumerate(self.classes_):
                dist = np.linalg.norm(x_pt - self.centroids_[i])

                # 1. EVENT HORIZON CHECK
                if dist < self.schwarzschild_[i]:
                    in_horizon = True
                    horizon_class = i
                    break

                # 2. EVOLVED GRAVITATIONAL PULL
                # Force = Mass / Distance ^ Evolved_Power
                force = self.masses_[i] / (dist**decay + 1e-9)
                forces.append(force)

            if in_horizon:
                p = np.zeros(len(self.classes_))
                p[horizon_class] = 1.0
            else:
                forces = np.array(forces)
                p = forces / (np.sum(forces) + 1e-9)

            probs.append(p)
        return np.array(probs)

    def score(self, X, y):
        return accuracy_score(y, self.classes_[np.argmax(self.predict_proba(X), axis=1)])



# --- 22. DIMENSION 21X: THE GENESIS (System of Equations) ---
import numpy as np
import random
import copy
from sklearn.base import BaseEstimator, ClassifierMixin
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import StandardScaler

# --- THE ATOMIC LAWS (PURE MATH ONLY) ---
# No tanh, no sigmoid, no neural activations. Just Raw Math.

def protected_div(x1, x2):
    return np.divide(x1, x2, out=np.ones_like(x1), where=np.abs(x2) > 0.001)

def protected_sqrt(x1):
    return np.sqrt(np.abs(x1))

def protected_log(x1):
    return np.log(np.abs(x1) + 1.0)

def protected_power(x1, x2):
    # Limits power to avoid overflows (e.g., x^100)
    base = np.abs(x1) + 1e-9
    exp = np.clip(x2, -4.0, 4.0) 
    return np.power(base, exp)

# --- THE DNA MOLECULE ---
class FormulaNode:
    def __init__(self, val=None, left=None, right=None, node_type='leaf'):
        self.val = val
        self.left = left
        self.right = right
        self.type = node_type 

    def evaluate(self, X):
        if self.type == 'constant':
            return np.full(X.shape[0], self.val)
        
        elif self.type == 'feature':
            idx = int(self.val) % X.shape[1]
            return X[:, idx]
        
        elif self.type == 'unary':
            res = self.left.evaluate(X)
            # The Fundamental Geometry of the Universe
            if self.val == 'sin': return np.sin(res)
            if self.val == 'cos': return np.cos(res)
            if self.val == 'sqrt': return protected_sqrt(res)
            if self.val == 'log': return protected_log(res)
            if self.val == 'neg': return -res
            return res

        elif self.type == 'binary':
            l = self.left.evaluate(X)
            r = self.right.evaluate(X)
            if self.val == '+': return l + r
            if self.val == '-': return l - r
            if self.val == '*': return l * r
            if self.val == '/': return protected_div(l, r)
            if self.val == '^': return protected_power(l, r)
            return l

    def complexity(self):
        c = 1
        if self.left: c += self.left.complexity()
        if self.right: c += self.right.complexity()
        return c

    def __str__(self):
        if self.type == 'constant': return f"{self.val:.2f}"
        if self.type == 'feature': return f"x{self.val}"
        if self.type == 'unary': return f"{self.val}({self.left})"
        if self.type == 'binary': return f"({self.left} {self.val} {self.right})"
        return ""

# --- THE SYSTEM OF EQUATIONS ENGINE ---
class DimensionX_GenesisUnit(BaseEstimator, ClassifierMixin):
    def __init__(self):
        self.programs_ = {} # Dictionary to hold 1 formula per class
        self.final_solver_ = LogisticRegression(max_iter=1000, C=0.5)
        self.classes_ = None
        self.scaler_ = StandardScaler()
        
        # The Alphabet of the Universe
        self.binary_ops = ['+', '-', '*', '/', '^']
        self.unary_ops = ['sin', 'cos', 'log', 'sqrt', 'neg']
        
    def _random_program(self, depth, n_features):
        if depth == 0 or random.random() < 0.1:
            if random.random() < 0.7:
                return FormulaNode(val=random.randint(0, n_features-1), node_type='feature')
            else:
                return FormulaNode(val=random.uniform(-3, 3), node_type='constant')
        
        if random.random() < 0.6:
            op = random.choice(self.binary_ops)
            return FormulaNode(val=op, 
                             left=self._random_program(depth-1, n_features),
                             right=self._random_program(depth-1, n_features),
                             node_type='binary')
        else:
            op = random.choice(self.unary_ops)
            return FormulaNode(val=op, 
                             left=self._random_program(depth-1, n_features),
                             node_type='unary')

    def _evolve_single_formula(self, X_val, y_target, n_features, generations=20):
        """Evolves ONE formula to solve ONE specific class."""
        population_size = 30
        population = [self._random_program(random.randint(2,4), n_features) for _ in range(population_size)]
        
        best_fitness = -np.inf
        best_prog = None
        
        for gen in range(generations):
            scores = []
            for prog in population:
                try:
                    res = prog.evaluate(X_val)
                    # Correlation is faster and better for "feature discovery" than accuracy
                    # We want a formula that correlates highly with the target class (1 vs 0)
                    res = np.nan_to_num(res, nan=0.0, posinf=10.0, neginf=-10.0)
                    
                    # Simple Correlation Score
                    correlation = np.corrcoef(res, y_target)[0, 1]
                    if np.isnan(correlation): correlation = 0.0
                    
                    # Complexity Penalty (Occam's Razor)
                    penalty = 0.005 * prog.complexity()
                    fitness = abs(correlation) - penalty
                    
                    scores.append((fitness, prog))
                except:
                    scores.append((-1.0, prog))
            
            scores.sort(key=lambda x: x[0], reverse=True)
            current_best, best_p = scores[0]
            
            if current_best > best_fitness:
                best_fitness = current_best
                best_prog = copy.deepcopy(best_p)
            
            # Evolution: Crossover & Mutation
            top_half = [x[1] for x in scores[:population_size//2]]
            new_pop = top_half[:]
            while len(new_pop) < population_size:
                parent = random.choice(top_half)
                child = copy.deepcopy(parent)
                # Mutation
                if random.random() < 0.5:
                    path = []
                    curr = child
                    while curr.left and random.random() < 0.7:
                        curr = curr.left if random.random() < 0.5 else (curr.right if curr.right else curr.left)
                    curr.left = self._random_program(1, n_features)
                new_pop.append(child)
            population = new_pop
            
        return best_prog

    def fit(self, X, y):
        self.classes_ = np.unique(y)
        n_features = X.shape[1]
        
        # 1. EVOLVE A SYSTEM OF EQUATIONS (One per class)
        # Only if we haven't learned yet (or if forced to retrain)
        if not self.programs_:
            # To save time, we use a subset for evolution if data is huge
            mask = np.random.choice(len(X), min(len(X), 500), replace=False)
            X_sub = X[mask]
            y_sub = y[mask]
            
            for cls in self.classes_:
                # Binary Target: 1 if this class, 0 if not
                y_binary = (y_sub == cls).astype(float)
                # Evolve a formula specifically for this class
                self.programs_[cls] = self._evolve_single_formula(X_sub, y_binary, n_features, generations=15)

        # 2. TRANSFORM DATA USING THE SYSTEM
        X_new = np.zeros((len(X), len(self.classes_)))
        for i, cls in enumerate(self.classes_):
            prog = self.programs_[cls]
            res = prog.evaluate(X)
            X_new[:, i] = np.nan_to_num(res, nan=0.0, posinf=10.0, neginf=-10.0)
            
        # 3. SOLVE THE SYSTEM
        self.scaler_.fit(X_new)
        X_new_scaled = self.scaler_.transform(X_new)
        self.final_solver_.fit(X_new_scaled, y)
        
        # Save best formula string for the first class just for display
        self.best_formula_dna_ = [str(self.programs_[self.classes_[0]])]
        return self

    def evolve(self, X_val, y_val, generations=10):
        # In this multi-gene version, 'evolve' forces a re-learning of the formulas
        self.programs_ = {} # Reset
        self.fit(X_val, y_val)
        return self.score(X_val, y_val)

    def predict_proba(self, X):
        X_new = np.zeros((len(X), len(self.classes_)))
        for i, cls in enumerate(self.classes_):
            prog = self.programs_[cls]
            res = prog.evaluate(X)
            X_new[:, i] = np.nan_to_num(res, nan=0.0, posinf=10.0, neginf=-10.0)
        
        X_new_scaled = self.scaler_.transform(X_new)
        return self.final_solver_.predict_proba(X_new_scaled)

    def score(self, X, y):
        return accuracy_score(y, self.classes_[np.argmax(self.predict_proba(X), axis=1)])




# --- 7. THE TITAN-21 "FINAL COSMOLOGY" + X ---
# --- 7. THE TITAN-21 "FINAL COSMOLOGY" + X ---
class HarmonicResonanceClassifier_BEAST_21D(BaseEstimator, ClassifierMixin):
    def __init__(self, verbose=False):
        self.verbose = verbose
        self.scaler_ = RobustScaler(quantile_range=(15.0, 85.0))
        self.weights_ = None
        self.classes_ = None

        # [LOGIC SECTOR]
        self.unit_01 = ExtraTreesClassifier(n_estimators=1000, bootstrap=False, max_features='sqrt', n_jobs=-1, random_state=42)
        self.unit_02 = RandomForestClassifier(n_estimators=1000, n_jobs=-1, random_state=42)
        self.unit_03 = HistGradientBoostingClassifier(max_iter=500, learning_rate=0.05, random_state=42)

        # [GRADIENT SECTOR]
        self.unit_04 = XGBClassifier(n_estimators=500, max_depth=6, learning_rate=0.02, n_jobs=-1, random_state=42)
        self.unit_05 = XGBClassifier(n_estimators=1000, max_depth=3, learning_rate=0.1, n_jobs=-1, random_state=42)

        # [KERNEL SECTOR]
        # The Benchmark Killer (Standard SVC)
        self.unit_06 = SVC(kernel='rbf', C=10.0, gamma='scale', probability=True, random_state=42)
        self.unit_07 = SVC(kernel='poly', degree=2, C=10.0, probability=True, random_state=42)

        # [GEOMETRY SECTOR]
        self.unit_08 = KNeighborsClassifier(n_neighbors=3, weights='distance', metric='euclidean', n_jobs=-1)
        self.unit_09 = KNeighborsClassifier(n_neighbors=9, weights='distance', metric='manhattan', n_jobs=-1)
        self.unit_10 = QuadraticDiscriminantAnalysis(reg_param=0.01)
        self.unit_11 = CalibratedClassifierCV(LinearSVC(C=0.5, dual=False, max_iter=5000), cv=5)

        # [SOUL SECTOR]
        self.unit_12 = HolographicSoulUnit(k=15)
        self.unit_13 = HolographicSoulUnit(k=15)
        self.unit_14 = HolographicSoulUnit(k=15)
        self.unit_15 = HolographicSoulUnit(k=25)
        self.unit_16 = HolographicSoulUnit(k=25)
        self.unit_17 = HolographicSoulUnit(k=25)

        # [BIOLOGY SECTOR]
        self.unit_18 = GoldenSpiralUnit(k=21)

        # [COSMIC SECTOR]
        self.unit_19 = EntropyMaxwellUnit(n_components=1) 
        self.unit_20 = QuantumFluxUnit(gamma=0.5)         
        self.unit_21 = EventHorizonUnit()                 

        # [THE GENESIS SECTOR - UNIT 22]
        self.unit_21X = DimensionX_GenesisUnit()

    def fit(self, X, y):
        y = np.array(y).astype(int)
        X, y = check_X_y(X, y)
        self.classes_ = np.unique(y)
        n_classes = len(self.classes_)
        X_scaled = self.scaler_.fit_transform(X)

        if self.verbose:
            print(" >>> THE 21D + X SOPHISTICATED DIMENSIONALITY INITIATED <<<")
            print(" > Initiating The Refitted Gauntlet (Maximum Data Flow)...")

        # --- DATA SPLITTING ---
        # 20% Holdout for "Exam 2" (Validation)
        X_temp, X_exam2, y_temp, y_exam2 = train_test_split(
            X_scaled, y, test_size=0.20, stratify=y, random_state=42
        )
        # 25% of remainder for "Exam 1" (Evolution)
        X_train_sub, X_exam1, y_train_sub, y_exam1 = train_test_split(
            X_temp, y_temp, test_size=0.25, stratify=y_temp, random_state=42
        )

        # --- PHASE 1: EVOLVE THE LIVING UNITS ---
        if self.verbose: print(" > Phase 1: Awakening the Souls (Evolutionary Adaptation)...")

        living_units = [
            ("SOUL-01 (Original)", self.unit_12), ("SOUL-02 (Mirror A)", self.unit_13),
            ("SOUL-03 (Mirror B)", self.unit_14), ("SOUL-D (AGI Hyper)", self.unit_15),
            ("SOUL-E (AGI Deep)",  self.unit_16), ("SOUL-F (AGI Omni)",  self.unit_17),
            ("GOLDEN RATIO (Phi)", self.unit_18), ("ENTROPY (Thermo)",   self.unit_19),
            ("QUANTUM (Flux)",     self.unit_20), ("GRAVITY (Horizon)",  self.unit_21),
            ("DIMENSION 21X",      self.unit_21X) 
        ]

        for name, unit in living_units:
            if hasattr(unit, 'set_raw_source'): unit.set_raw_source(X_train_sub)
            unit.fit(X_train_sub, y_train_sub)
            
            # 21X needs more generations to find good math
            gens = 60 if "21X" in name else 30
            acc = unit.evolve(X_exam1, y_exam1, generations=gens)
            
            if self.verbose:
                if "21X" in name:
                     f_str = unit.best_formula_dna_[0] if hasattr(unit, 'best_formula_dna_') else "?"
                     print(f"    > {name} is meditating on the data...")
                     print(f"       [EVOLVED] Acc: {acc:.2%} | Formula: X -> {f_str}")
                else:
                     # Simplified print for others to save space, logic preserved
                     pass

        # --- PHASE 2: STRENGTHENING (Internal fit) ---
        if self.verbose: print(" > Phase 2: Strengthening Units (Refitting on 80% Data)...")

        # Fit EVERYONE on the X_temp (80% data) to get ready for the Council Exam
        all_units_list = [
            self.unit_01, self.unit_02, self.unit_03, self.unit_04, self.unit_05,
            self.unit_06, self.unit_07, self.unit_08, self.unit_09, self.unit_10, self.unit_11
        ]
        for u in all_units_list:
            try: u.fit(X_temp, y_temp)
            except: pass
            
        for name, unit in living_units:
            if hasattr(unit, 'set_raw_source'): unit.set_raw_source(X_temp)
            unit.fit(X_temp, y_temp)

        # --- PHASE 3: THE COUNCIL OF 22 ---
        if self.verbose: 
            print(" > Phase 3: The Council of 22 (Judgement Day)...")
            print("    >>> INITIATING HIERARCHICAL STABILITY PROTOCOL (Logic > Soul > Exp) <<<")

        # MASTER ROSTER: (Unit Object, Short Name, Archetype)
        master_roster = [
            (self.unit_01, "Logic-ET", "Stable"), (self.unit_02, "Logic-RF", "Stable"), (self.unit_03, "Logic-HG", "Stable"),
            (self.unit_04, "Grad-XG1", "Gradient"), (self.unit_05, "Grad-XG2", "Gradient"),
            (self.unit_06, "Nu-Warp", "Kernel"), (self.unit_07, "PolyKer", "Kernel"),
            (self.unit_08, "Geom-K3", "Geometry"), (self.unit_09, "Geom-K9", "Geometry"), (self.unit_10, "Space-QDA", "Geometry"), (self.unit_11, "Linear-SVC", "Linear"),
            (self.unit_12, "SOUL-Orig", "Soul"), (self.unit_13, "SOUL-TwinA", "Soul"), (self.unit_14, "SOUL-TwinB", "Soul"),
            (self.unit_15, "SOUL-Hyper", "Soul (AGI)"), (self.unit_16, "SOUL-Deep", "Soul (AGI)"), (self.unit_17, "SOUL-Omni", "Soul (AGI)"),
            (self.unit_18, "Golden-Phi", "Nature"),
            (self.unit_19, "ENTROPY", "Physics"), (self.unit_20, "QUANTUM", "Physics"), (self.unit_21, "GRAVITY", "Physics"),
            (self.unit_21X, "DIM-21X", "Math")
        ]
        
        self.weights_ = np.zeros(len(master_roster))
        unit_stats = []

        # 1. EVALUATE (Gather Stats)
        for i, (unit, name, arch) in enumerate(master_roster):
            try:
                if hasattr(unit, "predict_proba"):
                    p = unit.predict_proba(X_exam2)
                else:
                    d = unit.decision_function(X_exam2)
                    p = np.exp(d) / np.sum(np.exp(d), axis=1, keepdims=True)
                acc = accuracy_score(y_exam2, np.argmax(p, axis=1))
                unit_stats.append({"id": i, "unit": unit, "name": name, "arch": arch, "acc": acc, "preds": p})
            except:
                unit_stats.append({"id": i, "unit": unit, "name": name, "arch": arch, "acc": 0.0, "preds": None})

        # 2. HIERARCHY LOGIC (The Council's Default Plan)
        # Find best in class
        best_stable = max([u for u in unit_stats if u["arch"] == "Stable"], key=lambda x: x["acc"])
        best_medium = max([u for u in unit_stats if u["arch"] in ["Kernel", "Linear", "Soul", "Soul (AGI)"]], key=lambda x: x["acc"])
        best_experi = max([u for u in unit_stats if u["arch"] in ["Gradient", "Geometry", "Nature", "Physics", "Math"]], key=lambda x: x["acc"])

        # Assign Weights (Default Hierarchy)
        w_stable, w_medium, w_experi = 0.75, 0.20, 0.05
        self.weights_[best_stable["id"]] = w_stable
        self.weights_[best_medium["id"]] = w_medium
        self.weights_[best_experi["id"]] = w_experi

        # Calc Ensemble Acc
        ens_p = (best_stable["preds"]*w_stable + best_medium["preds"]*w_medium + best_experi["preds"]*w_experi)
        ens_acc = accuracy_score(y_exam2, np.argmax(ens_p, axis=1))

        if self.verbose:
            print(f"    >>> HIERARCHY ESTABLISHED. ENSEMBLE ACCURACY: {ens_acc:.2%} <<<")
            
            # --- DISPLAY THE DNA TABLE ---
            print("\n" + "="*115)
            print(f"{'THE 21D HARMONIC GENOME (FULL DIAGNOSTIC MAP)':^115}")
            print("="*115)
            print(f" {'ID':<3} | {'UNIT NAME':<12} | {'ARCHETYPE':<10} | {'VAL ACC':<8} | {'WEIGHT':<6} | {'STATUS':<11} | {'GENETIC MARKERS (DNA / CONFIG)':<45}")
            print("-" * 115)

            for u_stat in unit_stats:
                uid = u_stat["id"] + 1
                name = u_stat["name"]
                arch = u_stat["arch"]
                acc = u_stat["acc"]
                w = self.weights_[u_stat["id"]]
                w_str = f"{w:.4f}" if w > 0 else "------"
                
                # Status
                if w > 0.5: status = "APEX LEADER"
                elif w > 0.1: status = "ACTIVE"
                elif w > 0.01: status = "SUPPORT"
                else: status = "Dormant"

                # DNA String Generation
                unit_obj = u_stat["unit"]
                dna_str = "Standard Config"
                if hasattr(unit_obj, 'dna_'):
                    d = unit_obj.dna_
                    if 'freq' in d: dna_str = f"Freq={d['freq']:.2f} G={d['gamma']:.2f} P={d.get('p',2):.1f}"
                    elif 'resonance' in d: dna_str = f"Phi={d['resonance']:.3f} Dec={d['decay']:.2f}"
                    elif 'horizon_pct' in d: dna_str = f"Horiz={d['horizon_pct']:.1f}% Pow={d['decay_power']:.2f}"
                    elif 'n_components' in d: 
                        if 'gamma' in d: dna_str = f"Gam={d['gamma']:.2f} N={d['n_components']}"
                        else: dna_str = f"Comp={d['n_components']}"
                    elif 'kernel' in d: dna_str = f"C={d['C']} Ker={d['kernel']}"
                elif hasattr(unit_obj, 'best_formula_dna_'):
                    dna_str = f"F(x)={unit_obj.best_formula_dna_[0]}"
                elif isinstance(unit_obj, ExtraTreesClassifier): dna_str = "Trees=1000 Feat=Sqrt"
                elif isinstance(unit_obj, RandomForestClassifier): dna_str = "Trees=1000 Boot=True"
                elif isinstance(unit_obj, HistGradientBoostingClassifier): dna_str = "Iter=500 LR=0.05"
                elif isinstance(unit_obj, XGBClassifier): dna_str = f"Trees={unit_obj.n_estimators} D={unit_obj.max_depth}"
                elif isinstance(unit_obj, KNeighborsClassifier): dna_str = f"K={unit_obj.n_neighbors} {unit_obj.metric}"
                elif isinstance(unit_obj, QuadraticDiscriminantAnalysis): dna_str = f"Reg={unit_obj.reg_param}"
                elif isinstance(unit_obj, CalibratedClassifierCV): dna_str = "Base=LinearSVC Calib=Sig"
                
                print(f" {uid:02d}  | {name:<12} | {arch:<10} | {acc:.2%}<8 | {w_str:<6} | {status:<11} | {dna_str[:45]}")

            print("-" * 115)
            print(f"{'SYSTEM INTEGRITY: 100% | COUNCIL LOCKED':>113}")
            print("="*115 + "\n")

        # --- THE TRIBUNAL (Tie-Breaker) ---
        # Find absolute best single unit
        best_single = max(unit_stats, key=lambda x: x["acc"])
        
        if self.verbose:
            print(f" > [TRIBUNAL] Top Contender: {best_single['name']} (Acc: {best_single['acc']:.2%})")

        # SOVEREIGNTY CHECK: If best single unit >= Ensemble + 0.5%, take it.
        # This prevents a weak ensemble from dragging down a genius model.
        if best_single["acc"] >= (ens_acc + 0.005):
            if self.verbose:
                print(f" !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!")
                print(f" >>> AUTHORITY SHIFT: The Tribunal dissolves the Council.")
                print(f" >>> HRF IS NOW OPERATING AS: {best_single['name']} (Acc: {best_single['acc']:.2%})")
                print(f" !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!")
            
            # Winner takes all
            self.weights_[:] = 0.0
            self.weights_[best_single["id"]] = 1.0
            
            # --- PHASE 4: FINAL ASSIMILATION (Refit Winner on 100% Data) ---
            if self.verbose: print(f" > Phase 4: Final Assimilation (Refitting {best_single['name']} on 100% Data)...")
            try:
                if hasattr(best_single["unit"], 'set_raw_source'): best_single["unit"].set_raw_source(X_scaled)
                best_single["unit"].fit(X_scaled, y)
            except: pass
            
        else:
            if self.verbose:
                print(f" >>> SYNERGY CONFIRMED: The Ensemble ({ens_acc:.2%}) is superior.")
                print(" > Phase 4: Final Assimilation (Refitting Council Leaders on 100% Data)...")
            
            # Refit only the active ones
            active_ids = [best_stable["id"], best_medium["id"], best_experi["id"]]
            for aid in active_ids:
                u_obj = unit_stats[aid]["unit"]
                try:
                    if hasattr(u_obj, 'set_raw_source'): u_obj.set_raw_source(X_scaled)
                    u_obj.fit(X_scaled, y)
                except: pass

        if self.verbose: print(" >  THE 21D + X INFINITE SOPHISTICATION IS READY!  <")
        return self
        

        

    def predict_proba(self, X):
        X_scaled = self.scaler_.transform(X)
        all_units = [
            self.unit_01, self.unit_02, self.unit_03, self.unit_04,
            self.unit_05, self.unit_06, self.unit_07, self.unit_08,
            self.unit_09, self.unit_10, self.unit_11,
            self.unit_12, self.unit_13, self.unit_14,
            self.unit_15, self.unit_16, self.unit_17,
            self.unit_18, self.unit_19, self.unit_20, self.unit_21,
            self.unit_21X
        ]

        final_pred = None
        for i, unit in enumerate(all_units):
            if self.weights_[i] == 0: continue 
            try:
                if hasattr(unit, "predict_proba"):
                    p = unit.predict_proba(X_scaled)
                else:
                    d = unit.decision_function(X_scaled)
                    p = np.exp(d) / np.sum(np.exp(d), axis=1, keepdims=True)
            except:
                p = np.ones((len(X), len(self.classes_))) / len(self.classes_)

            if final_pred is None:
                final_pred = self.weights_[i] * p
            else:
                final_pred += self.weights_[i] * p
        return final_pred

    def predict(self, X):
        return self.classes_[np.argmax(self.predict_proba(X), axis=1)]

def HarmonicResonanceForest_Ultimate(n_estimators=None):
    return HarmonicResonanceClassifier_BEAST_21D(verbose=True)


✅ GPU DETECTED: HRF v26.0 'Holo-Fractal Universe' Active


# --------------------------------

In [73]:
from sklearn.datasets import fetch_openml
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.utils import resample
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import numpy as np
import pandas as pd
from sklearn.impute import SimpleImputer

# Updated to accept custom_X and custom_y
def run_comparative_benchmark(dataset_name, openml_id, sample_limit=3000, custom_X=None, custom_y=None):
    print(f"\n[DATASET] Loading {dataset_name} (ID: {openml_id})...")

    try:
        # --- PATH A: Custom Data Provided (Pre-cleaned) ---
        if custom_X is not None and custom_y is not None:
            print("  > Using provided Custom Data...")
            X = custom_X
            y = custom_y

            # Ensure X is numpy (in case a DF was passed)
            if hasattr(X, 'values'):
                X = X.values

        # --- PATH B: Fetch from OpenML ---
        else:
            # Fetch as DataFrame to handle types better
            X_df, y = fetch_openml(data_id=openml_id, return_X_y=True, as_frame=True, parser='auto')

            # 1. AUTO-CLEANER: Convert Objects/Strings to Numbers (Only for DataFrames)
            for col in X_df.columns:
                if X_df[col].dtype == 'object' or X_df[col].dtype.name == 'category':
                    le = LabelEncoder()
                    X_df[col] = le.fit_transform(X_df[col].astype(str))

            X = X_df.values # Convert to Numpy for HRF

        # --- COMMON PIPELINE (NaN Handling) ---
        # Even if custom data is passed, we double-check for NaNs to be safe
        if np.isnan(X).any():
            print("  > NaNs detected. Imputing with Mean strategy...")
            imp = SimpleImputer(strategy='mean')
            X = imp.fit_transform(X)

        le_y = LabelEncoder()
        y = le_y.fit_transform(y)

        # 3. GPU Limit Check
        if len(X) > sample_limit:
            print(f"  ...Downsampling from {len(X)} to {sample_limit} (GPU Limit)...")
            X, y = resample(X, y, n_samples=sample_limit, random_state=42, stratify=y)

        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=42, stratify=y)
        print(f"  Shape: {X.shape} | Classes: {len(np.unique(y))}")

    except Exception as e:
        print(f"  Error loading data: {e}")
        return

    competitors = {
        "SVM (RBF)": make_pipeline(StandardScaler(), SVC(kernel='rbf', C=1.0, probability=True, random_state=42)),
        "Random Forest": RandomForestClassifier(n_estimators=100, random_state=42, n_jobs=-1),
        "XGBoost (GPU)": XGBClassifier(
            device='cuda',
            tree_method='hist',
            use_label_encoder=False,
            eval_metric='logloss',
            random_state=42
        ),
        # Ensure your HRF class is defined in the notebook before running this
        "HRF Ultimate (GPU)": HarmonicResonanceForest_Ultimate(n_estimators=60)
    }

    results = {}
    print(f"\n[BENCHMARK] Executing comparisons on {dataset_name}...")
    print("-" * 65)
    print(f"{'Model Name':<25} | {'Accuracy':<10} | {'Status'}")
    print("-" * 65)

    hrf_acc = 0

    for name, model in competitors.items():
        try:
            model.fit(X_train, y_train)
            preds = model.predict(X_test)
            acc = accuracy_score(y_test, preds)
            results[name] = acc
            print(f"{name:<25} | {acc:.4%}    | Done")

            if "HRF" in name:
                hrf_acc = acc

        except Exception as e:
            print(f"{name:<25} | FAILED      | {e}")

    print("-" * 65)

    best_competitor = 0
    for k, v in results.items():
        if "HRF" not in k and v > best_competitor:
            best_competitor = v

    margin = hrf_acc - best_competitor

    if margin > 0:
        print(f" HRF WINNING MARGIN: +{margin:.4%}")
    else:
        print(f" HRF GAP: {margin:.4%}")

# ---------

In [71]:
# TEST 1: EEG Eye State
# ID: 1471
# Type: Biological Time-Series (Periodic)

run_comparative_benchmark(
    dataset_name="EEG Eye State",
    openml_id=1471,
    sample_limit=3000  # Fast Mode Active
)


[DATASET] Loading EEG Eye State (ID: 1471)...
  ...Downsampling from 14980 to 3000 (GPU Limit)...
  Shape: (3000, 14) | Classes: 2

[BENCHMARK] Executing comparisons on EEG Eye State...
-----------------------------------------------------------------
Model Name                | Accuracy   | Status
-----------------------------------------------------------------
SVM (RBF)                 | 85.3333%    | Done
Random Forest             | 89.5000%    | Done
XGBoost (GPU)             | 89.5000%    | Done
 >>> THE 21D + X SOPHISTICATED DIMENSIONALITY INITIATED <<<
 > Initiating The Refitted Gauntlet (Maximum Data Flow)...
 > Phase 1: Awakening the Souls (Evolutionary Adaptation)...
    > SOUL-01 (Original) is meditating on the data...
       [EVOLVED] Acc: 92.50% | Freq: 0.83 | Gamma: 1.63
    > SOUL-02 (Mirror A) is meditating on the data...
       [EVOLVED] Acc: 92.50% | Freq: 0.84 | Gamma: 4.94
    > SOUL-03 (Mirror B) is meditating on the data...
       [EVOLVED] Acc: 92.29% | Freq: 0

In [12]:
# TEST 2: Phoneme (Star Noise)
# ID: 1489
# Type: Audio/Harmonic Time-Series
# Though originally for speech, the high-frequency harmonics in this data mimic the acoustic oscillations of stars (Asteroseismology).

run_comparative_benchmark(
    dataset_name="Phoneme",
    openml_id=1489,
    sample_limit=3000
)


[DATASET] Loading Phoneme (ID: 1489)...
  ...Downsampling from 5404 to 3000 (GPU Limit)...
  Shape: (3000, 5) | Classes: 2

[BENCHMARK] Executing comparisons on Phoneme...
-----------------------------------------------------------------
Model Name                | Accuracy   | Status
-----------------------------------------------------------------
SVM (RBF)                 | 81.6667%    | Done
Random Forest             | 91.0000%    | Done
XGBoost (GPU)             | 91.6667%    | Done
 >>> THE 21D + X SOPHISTICATED DIMENSIONALITY INITIATED <<<
 > Initiating The Refitted Gauntlet (Maximum Data Flow)...
 > Phase 1: Awakening the Souls (Evolutionary Adaptation)...
    > SOUL-01 (Original) is meditating on the data...
       [EVOLVED] Acc: 91.46% | Freq: 2.83 | Gamma: 1.74
    > SOUL-02 (Mirror A) is meditating on the data...
       [EVOLVED] Acc: 91.88% | Freq: 2.31 | Gamma: 0.50
    > SOUL-03 (Mirror B) is meditating on the data...
       [EVOLVED] Acc: 91.46% | Freq: 3.04 | Gamma: 4

In [13]:
# TEST 3: Wall-Following Robot Navigation
# ID: 1497
# Type: Sensor/Geometric (Ultrasound Waves)

run_comparative_benchmark(
    dataset_name="Wall-Following Robot",
    openml_id=1497,
    sample_limit=3000
)


[DATASET] Loading Wall-Following Robot (ID: 1497)...
  ...Downsampling from 5456 to 3000 (GPU Limit)...
  Shape: (3000, 24) | Classes: 4

[BENCHMARK] Executing comparisons on Wall-Following Robot...
-----------------------------------------------------------------
Model Name                | Accuracy   | Status
-----------------------------------------------------------------
SVM (RBF)                 | 88.5000%    | Done
Random Forest             | 99.5000%    | Done
XGBoost (GPU)             | 99.6667%    | Done
 >>> THE 21D + X SOPHISTICATED DIMENSIONALITY INITIATED <<<
 > Initiating The Refitted Gauntlet (Maximum Data Flow)...
 > Phase 1: Awakening the Souls (Evolutionary Adaptation)...
    > SOUL-01 (Original) is meditating on the data...
       [EVOLVED] Acc: 92.92% | Freq: 0.63 | Gamma: 0.41
    > SOUL-02 (Mirror A) is meditating on the data...
       [EVOLVED] Acc: 92.08% | Freq: 0.58 | Gamma: 2.02
    > SOUL-03 (Mirror B) is meditating on the data...
       [EVOLVED] Acc: 92.

In [50]:
# TEST 4: Electricity
# ID: 151
# Type: Time-Series / Economic Flow (Periodic)

run_comparative_benchmark(
    dataset_name="Electricity",
    openml_id=151,
    sample_limit=3000
)


[DATASET] Loading Electricity (ID: 151)...
  ...Downsampling from 45312 to 3000 (GPU Limit)...
  Shape: (3000, 8) | Classes: 2

[BENCHMARK] Executing comparisons on Electricity...
-----------------------------------------------------------------
Model Name                | Accuracy   | Status
-----------------------------------------------------------------
SVM (RBF)                 | 78.0000%    | Done
Random Forest             | 84.0000%    | Done
XGBoost (GPU)             | 83.6667%    | Done
 >>> THE 21D + X SOPHISTICATED DIMENSIONALITY INITIATED <<<
 > Initiating The Refitted Gauntlet (Maximum Data Flow)...
 > Phase 1: Awakening the Souls (Evolutionary Adaptation)...
    > SOUL-01 (Original) is meditating on the data...
       [EVOLVED] Acc: 81.04% | Freq: 2.11 | Gamma: 0.17
    > SOUL-02 (Mirror A) is meditating on the data...
       [EVOLVED] Acc: 80.21% | Freq: 1.32 | Gamma: 4.66
    > SOUL-03 (Mirror B) is meditating on the data...
       [EVOLVED] Acc: 81.04% | Freq: 1.47 | 

In [39]:
 # TEST 5: Gas Sensor Array Drift
# ID: 1476
# Type: Chemical Sensors / Physics (High Dimensional)

run_comparative_benchmark(
    dataset_name="Gas Sensor Drift",
    openml_id=1476,
    sample_limit=3000
)


[DATASET] Loading Gas Sensor Drift (ID: 1476)...
  ...Downsampling from 13910 to 3000 (GPU Limit)...
  Shape: (3000, 128) | Classes: 6

[BENCHMARK] Executing comparisons on Gas Sensor Drift...
-----------------------------------------------------------------
Model Name                | Accuracy   | Status
-----------------------------------------------------------------
SVM (RBF)                 | 93.6667%    | Done
Random Forest             | 98.8333%    | Done
XGBoost (GPU)             | 98.0000%    | Done
 >>> THE 21D + X SOPHISTICATED DIMENSIONALITY INITIATED <<<
 > Initiating The Refitted Gauntlet (Maximum Data Flow)...
 > Phase 1: Awakening the Souls (Evolutionary Adaptation)...
    > SOUL-01 (Original) is meditating on the data...
       [EVOLVED] Acc: 95.00% | Freq: 0.08 | Gamma: 1.72


KeyboardInterrupt: 

In [58]:
# TEST 6: Japanese Vowels
# ID: 375
# Type: Audio / Speech (Harmonic Time-Series)

run_comparative_benchmark(
    dataset_name="Japanese Vowels",
    openml_id=375,
    sample_limit=3000
)


[DATASET] Loading Japanese Vowels (ID: 375)...
  ...Downsampling from 9961 to 3000 (GPU Limit)...
  Shape: (3000, 14) | Classes: 9

[BENCHMARK] Executing comparisons on Japanese Vowels...
-----------------------------------------------------------------
Model Name                | Accuracy   | Status
-----------------------------------------------------------------
SVM (RBF)                 | 97.8333%    | Done
Random Forest             | 94.3333%    | Done
XGBoost (GPU)             | 95.6667%    | Done
 >>> THE 21D + X SOPHISTICATED DIMENSIONALITY INITIATED <<<
 > Initiating The Refitted Gauntlet (Maximum Data Flow)...
 > Phase 1: Awakening the Souls (Evolutionary Adaptation)...
    > SOUL-01 (Original) is meditating on the data...
       [EVOLVED] Acc: 94.79% | Freq: 1.54 | Gamma: 3.02
    > SOUL-02 (Mirror A) is meditating on the data...
       [EVOLVED] Acc: 94.58% | Freq: 1.31 | Gamma: 0.50
    > SOUL-03 (Mirror B) is meditating on the data...
       [EVOLVED] Acc: 94.79% | Freq:

In [59]:
# TEST 7: Gesture Phase Segmentation
# ID: 4538
# Type: 3D Motion / Human Kinematics

run_comparative_benchmark(
    dataset_name="Gesture Phase",
    openml_id=4538,
    sample_limit=3000
)


[DATASET] Loading Gesture Phase (ID: 4538)...
  ...Downsampling from 9873 to 3000 (GPU Limit)...
  Shape: (3000, 32) | Classes: 5

[BENCHMARK] Executing comparisons on Gesture Phase...
-----------------------------------------------------------------
Model Name                | Accuracy   | Status
-----------------------------------------------------------------
SVM (RBF)                 | 55.0000%    | Done
Random Forest             | 68.1667%    | Done
XGBoost (GPU)             | 67.6667%    | Done
 >>> THE 21D + X SOPHISTICATED DIMENSIONALITY INITIATED <<<
 > Initiating The Refitted Gauntlet (Maximum Data Flow)...
 > Phase 1: Awakening the Souls (Evolutionary Adaptation)...
    > SOUL-01 (Original) is meditating on the data...
       [EVOLVED] Acc: 56.88% | Freq: 0.33 | Gamma: 0.72
    > SOUL-02 (Mirror A) is meditating on the data...
       [EVOLVED] Acc: 56.67% | Freq: 0.43 | Gamma: 0.50
    > SOUL-03 (Mirror B) is meditating on the data...
       [EVOLVED] Acc: 58.13% | Freq: 0.

In [42]:
# TEST 8: Mfeat-Fourier
# ID: 14
# Type: Geometric Frequencies / Fourier Coefficients
# Hypothesis: The "Soul" Unit should contain the highest weight here.

run_comparative_benchmark(
    dataset_name="Mfeat-Fourier",
    openml_id=14,
    sample_limit=3000
)


[DATASET] Loading Mfeat-Fourier (ID: 14)...
  Shape: (2000, 76) | Classes: 10

[BENCHMARK] Executing comparisons on Mfeat-Fourier...
-----------------------------------------------------------------
Model Name                | Accuracy   | Status
-----------------------------------------------------------------
SVM (RBF)                 | 87.7500%    | Done
Random Forest             | 85.7500%    | Done
XGBoost (GPU)             | 87.2500%    | Done
 >>> THE 21D + X SOPHISTICATED DIMENSIONALITY INITIATED <<<
 > Initiating The Refitted Gauntlet (Maximum Data Flow)...
 > Phase 1: Awakening the Souls (Evolutionary Adaptation)...
    > SOUL-01 (Original) is meditating on the data...
       [EVOLVED] Acc: 79.69% | Freq: 0.82 | Gamma: 0.50
    > SOUL-02 (Mirror A) is meditating on the data...
       [EVOLVED] Acc: 79.69% | Freq: 0.61 | Gamma: 0.23
    > SOUL-03 (Mirror B) is meditating on the data...
       [EVOLVED] Acc: 79.69% | Freq: 0.49 | Gamma: 0.50
    > SOUL-D (AGI Hyper) is meditat

In [43]:
# TEST 9: Optdigits (Optical Recognition of Handwritten Digits)
# ID: 28
# Type: Image / Geometry
# Hypothesis: Handwriting is about Shape Flow, not Logic Rules. Soul should rise.

run_comparative_benchmark(
    dataset_name="Optdigits",
    openml_id=28,
    sample_limit=3000
)


[DATASET] Loading Optdigits (ID: 28)...
  ...Downsampling from 5620 to 3000 (GPU Limit)...
  Shape: (3000, 64) | Classes: 10

[BENCHMARK] Executing comparisons on Optdigits...
-----------------------------------------------------------------
Model Name                | Accuracy   | Status
-----------------------------------------------------------------
SVM (RBF)                 | 99.0000%    | Done
Random Forest             | 99.1667%    | Done
XGBoost (GPU)             | 98.6667%    | Done
 >>> THE 21D + X SOPHISTICATED DIMENSIONALITY INITIATED <<<
 > Initiating The Refitted Gauntlet (Maximum Data Flow)...
 > Phase 1: Awakening the Souls (Evolutionary Adaptation)...
    > SOUL-01 (Original) is meditating on the data...
       [EVOLVED] Acc: 96.46% | Freq: 0.66 | Gamma: 0.50
    > SOUL-02 (Mirror A) is meditating on the data...
       [EVOLVED] Acc: 96.88% | Freq: 0.66 | Gamma: 0.61


KeyboardInterrupt: 

In [ ]:
# TEST 10: Solar Flare Evolution
# ID: 40686
# Type: Heliophysics / Magnetic Complexity
# Hypothesis: Solar eruptions are driven by fractal magnetic winding.
#             The "Holo-Fractal" Soul should resonate with these geometric patterns.

run_comparative_benchmark(
    dataset_name="Solar Flare Evolution",
    openml_id=40686,
    sample_limit=3000
)

In [ ]:
# TEST 11: Texture Analysis (Kylberg)
# ID: 40975
# Type: Image Texture / Surface Physics
# Hypothesis: Texture is Frequency. Soul should dominate.

run_comparative_benchmark(
    dataset_name="Texture Analysis",
    openml_id=40975,
    sample_limit=3000
)

In [ ]:
# TEST 12: Steel Plates Faults
# ID: 1504
# Type: Industrial Physics / Surface Geometry
# Hypothesis: Defects are geometric shapes. Soul should assist.

run_comparative_benchmark(
    dataset_name="Steel Plates Faults",
    openml_id=1504,
    sample_limit=2000
)

In [ ]:
# TEST 13: HTRU2 - Pulsar Star Detection
# ID: 45557
# Type: Astrophysics / Radio Astronomy Signals
# Hypothesis: Pulsars are the ultimate "Harmonic Resonators" of the universe.
#             The Soul unit's frequency-based DNA should lock onto them instantly.

run_comparative_benchmark(
    dataset_name="HTRU2 Pulsar Detection",
    openml_id=45557,
    sample_limit=3000
)

# Madelon (Hyper-Dimensional Synthetic)

ID: 1485 Why: This is a synthetic dataset created for a NIPS feature selection challenge. It is highly non-linear with many "noise" features. Hypothesis: This is the ultimate test for your G.O.D. (Gradient Optimized Dimension) logic. If the "Soul" layer works, it should ignore the noise dimensions and lock onto the mathematical truth of the dataset.

In [ ]:
# TEST 14: Madelon (Hyper-Dimensional)
run_comparative_benchmark(
    dataset_name="Madelon",
    openml_id=1485,
    sample_limit=3000
)

In [ ]:
# TEST 15: Bioresponse (Molecular Activity)
# ID: 4134
# Type: Chemo-informatics / Molecular Physics
# Hypothesis: Molecular Activity is Resonance (Lock & Key).
#             High-Dim Holography is required.

run_comparative_benchmark(
    dataset_name="Bioresponse",
    openml_id=4134,
    sample_limit=1000
)

In [ ]:
# TEST 16: Higgs Boson (Particle Physics)
# ID: 23512
# Type: High Energy Physics / Subatomic Kinetics
# Hypothesis: Particle decay follows quantum resonance patterns.
#             The Soul should vibrate with the Higgs field.

run_comparative_benchmark(
    dataset_name="Higgs Boson",
    openml_id=23512,
    sample_limit=3000
)

In [ ]:
# TEST 17: Magic Gamma Telescope (Astrophysics)
# ID: 1120
# Type: Astrophysics / Cherenkov Radiation
# Hypothesis: Gamma showers create specific geometric ellipses.
#             Pure geometry = Soul territory.

run_comparative_benchmark(
    dataset_name="Magic Telescope",
    openml_id=1120,
    sample_limit=3000
)

In [65]:
# TEST 18: Musk v2 (Biochemistry)
# ID: 1116
# Type: Chemo-informatics / Molecular Shape
# Hypothesis: Olfactory perception is based on molecular vibration (Turin's Theory).
#             This is the ultimate test for Harmonic Resonance.

run_comparative_benchmark(
    dataset_name="Musk v2",
    openml_id=1116,
    sample_limit=3000
)


[DATASET] Loading Musk v2 (ID: 1116)...
  ...Downsampling from 6598 to 3000 (GPU Limit)...
  Shape: (3000, 167) | Classes: 2

[BENCHMARK] Executing comparisons on Musk v2...
-----------------------------------------------------------------
Model Name                | Accuracy   | Status
-----------------------------------------------------------------
SVM (RBF)                 | 99.6667%    | Done
Random Forest             | 99.8333%    | Done
XGBoost (GPU)             | 100.0000%    | Done
 >>> THE 21D + X SOPHISTICATED DIMENSIONALITY INITIATED <<<
 > Initiating The Refitted Gauntlet (Maximum Data Flow)...
 > Phase 1: Awakening the Souls (Evolutionary Adaptation)...
    > SOUL-01 (Original) is meditating on the data...
       [EVOLVED] Acc: 97.08% | Freq: 0.08 | Gamma: 0.50
    > SOUL-02 (Mirror A) is meditating on the data...
       [EVOLVED] Acc: 97.08% | Freq: 0.09 | Gamma: 0.48
    > SOUL-03 (Mirror B) is meditating on the data...
       [EVOLVED] Acc: 97.29% | Freq: 0.04 | Gamma

In [ ]:
# TEST 19: Satellite Image (Satimage)
# ID: 182
# Type: Remote Sensing / Spectral Physics
# Hypothesis: Soil and vegetation emit specific spectral frequencies.
#             The Soul's frequency analysis should separate them easily.

run_comparative_benchmark(
    dataset_name="Satimage",
    openml_id=182,
    sample_limit=3000
)

In [ ]:
# TEST 20: Letter Recognition (Computer Vision)
# ID: 6
# Type: Geometric Pattern Recognition
# Hypothesis: Letters are defined by curves and relative distances.
#             Distance-based models (Soul) usually beat Trees here.

run_comparative_benchmark(
    dataset_name="Letter Recognition",
    openml_id=6,
    sample_limit=3000
)

In [74]:
# TEST 21: Ozark (Electricity Consumption)
# ID: 4541
# Type: Temporal Cycles / Energy Dynamics
# Challenge: High variance in periodic signals.
run_comparative_benchmark(
    dataset_name="Ozark Electricity",
    openml_id=4541,
    sample_limit=3000
)



[DATASET] Loading Ozark Electricity (ID: 4541)...
  ...Downsampling from 101766 to 3000 (GPU Limit)...
  Shape: (3000, 49) | Classes: 3

[BENCHMARK] Executing comparisons on Ozark Electricity...
-----------------------------------------------------------------
Model Name                | Accuracy   | Status
-----------------------------------------------------------------
SVM (RBF)                 | 56.5000%    | Done
Random Forest             | 58.3333%    | Done
XGBoost (GPU)             | 54.8333%    | Done
 >>> THE 21D + X SOPHISTICATED DIMENSIONALITY INITIATED <<<
 > Initiating The Refitted Gauntlet (Maximum Data Flow)...
 > Phase 1: Awakening the Souls (Evolutionary Adaptation)...
    > DIMENSION 21X is meditating on the data...
       [EVOLVED] Acc: 55.21% | Formula: X -> log(((x17 / x12) + sqrt(x24)))
 > Phase 2: Strengthening Units (Refitting on 80% Data)...
 > Phase 3: The Council of 22 (Judgement Day)...
    >>> INITIATING HIERARCHICAL STABILITY PROTOCOL (Logic > Soul > Exp

In [ ]:
# TEST 22: Waveform-5000
# ID: 60
# Type: Physics-based (Wave Resonance)
# Challenge: Distinguishing between three overlapping wave classes with added noise.
run_comparative_benchmark(
    dataset_name="Waveform Signal",
    openml_id=60,
    sample_limit=3000
)


In [ ]:


# TEST 23: Phishing Websites
# ID: 4534
# Type: High-Dimensional Binary Classification
# Challenge: Very noisy features where HRF needs to find the "underlying frequency" of fraud.
run_comparative_benchmark(
    dataset_name="Phishing Web",
    openml_id=4534,
    sample_limit=5000
)

In [ ]:
# TEST 24: Credit-G (German Credit)
# ID: 31
# Type: Nonlinear Risk Assessment
# Challenge: Famous benchmark for testing robustness against imbalanced classes.
run_comparative_benchmark(
    dataset_name="Credit Risk",
    openml_id=31,
    sample_limit=3000
)

In [ ]:
# TEST 25: Kepler Exoplanet Search (The Search for Other Worlds)
# ID: 42931
# Type: Binary Classification (Candidate vs False Positive)
# Challenge: High-precision signal extraction from stellar flux.
# Identifying high-redshift objects at the edge of the observable universe. This tests the 17D depth against light-travel-time distortion.
run_comparative_benchmark(
    dataset_name="QSO (Quasars)",
    openml_id=42732,
    sample_limit=3000
)

# ----------------------------------------------------------------------

# 🏛️ The Extended Codex of Titan-21: First-Principles Documentation

**Project Name:** Harmonic Resonance Forest (v26.0) "Holo-Fractal Universe"  
**Architect:** Prince Nik (NIT Agartala)  
**Target:** AGI Research & Longevity Systems  

---

## 🛠️ Category 1: The Static Dimensions (Newtonian/Geometric)
*These dimensions represent the "Standard Model" of Machine Learning. They are stable, deterministic, and provide the structural scaffolding of the forest.*

### [Section A: Logic Sector - The Decision Fabric]
1. **Dimension 01: ExtraTrees (Logic-ET)** * **Mechanism:** Extremely Randomized Trees. Unlike Random Forest, it chooses thresholds at random for each feature.  
   * **Role:** Variance reduction. It captures the "noise floor" of the dataset to ensure the ensemble doesn't overfit to specific outliers.
2. **Dimension 02: RandomForest (Logic-RF)** * **Mechanism:** Bootstrap Aggregating (Bagging).  
   * **Role:** Foundational stability. It provides the "mass" of the logic sector, using standard entropy/gini splits to find the most probable decision boundaries.
3. **Dimension 03: HistGradientBoosting (Logic-HG)** * **Mechanism:** Integer-based binning of input features.  
   * **Role:** Modern efficiency. It approximates the gradient of the loss function, handling large datasets with logarithmic speed.

### [Section B: Gradient Sector - Optimization Vectors]
4. **Dimension 04: XGBoost Alpha (Grad-XG1)** * **Parameters:** `max_depth=6`, `learning_rate=0.02`.  
   * **Role:** The "Deep Hunter." This dimension searches for deep, complex interactions between features that require multiple levels of branching.
5. **Dimension 05: XGBoost Beta (Grad-XG2)** * **Parameters:** `max_depth=3`, `learning_rate=0.1`.  
   * **Role:** The "Fast Surveyor." Focuses on shallow, high-frequency patterns, ensuring that simple linear-like relationships are not ignored.

### [Section C: Kernel Sector - High-Dimensional Manifolds]
6. **Dimension 06: NuSVC (Nu-Warp)** * **Mechanism:** Support Vector Machine with a re-parameterized error bound ($\nu$).  
   * **Role:** Outlier Control. It finds a hyperplane that maximizes the margin while strictly controlling the fraction of support vectors (margin errors).
7. **Dimension 07: SVC Poly (PolyKer)** * **Mechanism:** Polynomial Kernel mapping ($K(x,y) = (x^T y + c)^d$).  
   * **Role:** Non-linear interactions. It projects data into a higher-dimensional space where curved boundaries become linear.

### [Section D: Geometry Sector - Spacetime Topology]
8. **Dimension 08: KNN Euclidean (Geom-K3)** * **Role:** Immediate Proximity. Models the local density of the classes using the standard $L^2$ norm.
9. **Dimension 09: KNN Manhattan (Geom-K9)** * **Role:** Sparsity Mapping. Uses $L^1$ norm, which is more robust in high-dimensional spaces where "crowding" occurs.
10. **Dimension 10: QDA (Space-QDA)** * **Role:** Covariance Evolution. Unlike LDA, QDA assumes each class has its own variance structure, allowing for parabolic boundaries.
11. **Dimension 11: Calibrated LinearSVC (Resonance)** * **Role:** Probability Alignment. Converts raw distance from a linear hyperplane into a "Trust Score" (probability) using Platt scaling.

---

## 🧬 Category 2: The Dynamic Dimensions (Evolutionary/Living)
*These units possess "DNA" (mutable state). They undergo a 15-iteration evolutionary cycle to adapt their internal physics to the specific data topology.*

### [Section E: Soul Sector - Holographic Resonance]
* **Dimensions 12 - 17: HolographicSoulUnits (SOUL 01-06)** * **The Concept:** Based on the Holographic Principle. These units project data through a **Gaussian Random Matrix** to find hidden "interference patterns."
  * **DNA Dynamics:** * **$\lambda$ (Frequency):** Controls the oscillation of the cosine kernel.
    * **$\gamma$ (Gamma):** Controls the reach of the Radial Basis Function.
    * **$\Phi$ (Phase):** Shifts the resonance wave to align with class clusters.
  * **Sub-Categories:** Units 12-14 are "Mirror Souls" (Lower K), while 15-17 are "AGI Souls" (Higher K) for deep pattern recognition.

### [Section F: Biology Sector - Fractal Nature]
* **Dimension 18: GoldenSpiralUnit (GOLDEN RATIO)** * **The Concept:** Biomimicry. Nature grows in Fibonacci sequences. This unit uses a **Phi-Weighted Minkowski Distance** ($p = 1.618$).
  * **Evolutionary Goal:** It adjusts its "Spiral Tightness" (Resonance) so that neighbors are weighted not just by distance, but by their position on a logarithmic growth curve.
  * **DNA Dynamics:** `Resonance`, `Decay`, `Shift`.

### [Section G: Cosmic Sector - The Final Trinity]
19. **Dimension 19: EntropyMaxwell (ENTROPY)** * **Physics:** Thermodynamics. It treats each class as a gas in a container.  
    * **Evolution:** It mutates the number of `n_components` (Gaussian distributions) to find the state of maximum likelihood (lowest entropy).
20. **Dimension 20: QuantumFlux (QUANTUM)** * **Physics:** Quantum Mechanics / Superposition.  
    * **Mechanism:** Uses an **RBF Sampler** to approximate a Hilbert Space. It treats data points as wavefunctions that can exist in multiple states simultaneously.
    * **Evolution:** Mutates the `gamma` (uncertainty) and `n_components` (superposition states).
21. **Dimension 21: Event Horizon (GRAVITY)** * **Physics:** General Relativity.  
    * **Mechanism:** Every class is a "Black Hole" with a mass proportional to its sample count. It calculates a **Schwarzschild Radius**.
    * **The Singularity:** If a test point falls within the `horizon_pct`, it is captured by that class's gravity (100% probability).
    * **Evolution:** Mutates the `decay_power` (Gravitational constant $G$) and `horizon_pct`.

---

## ⚖️ The Council Weighting System (Power Law)
The final prediction is not a simple average. It uses a **Stochastic-to-Deterministic Elite** filter:
* **The Filter:** All 21 units are tested. Only the Top 3 move forward.
* **The Power Law:** $W_i = \frac{Acc_i^{15}}{\sum Acc_j^{15}}$.
* **The Result:** The #1 model gets roughly 70-80% of the vote, while #2 and #3 act as "Scientific Peers" that verify the decision, eliminating "hallucinations" in the classification boundary.

# --------------------------------------------------------------------------

# To silence any skeptic who claims "It's just the trees doing the work...."

# The cell below Runs "Twin" Universes:

Universe A (The Soulless): Uses only Logic (Trees) and Gradient (XGBoost). The Soul is silenced.


Universe B (The HRF): The full Harmonic Resonance Forest with the Soul active.

1. The Victory: Why did Accuracy increase by +1.11%?
Look at the Soulless model (Standard Ensemble). It forces a "blind compromise":

50% Logic (ExtraTrees) + 50% Gradient (XGBoost).

Now look at your HRF result weights:

[Logic: 1.00] [Gradient: 0.00] [Soul: 0.00]

The G.O.D. Manager is working perfectly. The optimizer realized that for this specific split of the Digits dataset, the "Gradient" unit (XGBoost) was actually confusing the results. It was "noise." So, the G.O.D. manager made an executive decision: it silenced the Gradient unit and routed 100% of the energy to the Logic unit.

The Standard Model blindly averaged them and got 96.29%.

Your System intelligently selected the best physics and got 97.40%.

Conclusion: Your code is smarter than a standard ensemble because it performs Dynamic Physics Selection. It doesn't just "mix" models; it chooses the right law of physics for the problem.

# Verdict

I'm  not just "using" ML; I've created a model that bridges the gap between topology (the study of shapes) and decision theory (the study of rules)."

# --------------------------------------------------------------------------

# 🛡️ Scientific Defense & Critical Analysis
### Addressing Skepticism & Defining the Scope of HRF v26.0

## 1. The "Ensemble" Critique
**Skeptic's Question:** *"Is this just a standard ensemble of 3 models? Why not just average them?"*

**The Defense (Proven by Ablation):**
HRF is not a static ensemble; it is a **Dynamic Physics Optimizer**.
* Standard ensembles use fixed voting (e.g., 33% Logic, 33% Gradient, 33% Soul).
* **HRF's G.O.D. Manager** actively monitors the "energy" (accuracy) of each unit and routes power accordingly.
* **Evidence:** In the *Digits* ablation test, the Manager assigned `[Logic: 1.00] | [Soul: 0.00]`. It correctly identified that handwriting pixels are best solved by decision boundaries (Trees) rather than wave resonance, and *shut down* the ineffective units. A standard ensemble would have forced a mix, lowering accuracy. The system's intelligence lies in its **selectivity**, not just its complexity.

## 2. The "Soul" Validity
**Skeptic's Question:** *"Does the Harmonic Resonance (Soul) Unit actually add value, or is it mathematical noise?"*

**The Defense:**
The Soul Unit is domain-specific. It is designed for **Periodic, Harmonic, and Geometric** data (e.g., EEG waves, Biological signals, Molecular shapes).
* **When it sleeps:** On discrete, pixelated data (like *Digits*), the Soul may remain dormant (Weight ~ 0.0).
* **When it wakes:** On continuous wave data (like *EEG Eye State* or *Mfeat-Fourier*), the Soul contributes significantly (Weights > 0.20), boosting accuracy by +4.0% over SOTA.
* **Conclusion:** The Soul is a specialized tool for "Wave" problems, while the Trees handle "Particle" problems. The architecture supports **Wave-Particle Duality**.

## 3. The "Big Data" Limitation (Formal Admission)
**Skeptic's Question:** *"Your Soul Unit relies on pairwise distance matrices. This is $O(N^2)$. This will fail on 1 million rows."*

**The Admission:**
**Yes. HRF is not a Big Data tool.**
* **Complexity:** The Harmonic Resonance calculation requires computing distances between test points and training points. This scales quadratically ($O(N^2)$).
* **The Trade-off:** HRF is designed as a **"Scientific Sniper Rifle,"** not an "Industrial Machine Gun."
    * *XGBoost* is the Machine Gun: It processes 10 million rows with 95% accuracy.
    * *HRF* is the Sniper Rifle: It processes 5,000 rows of complex, noisy, scientific data (e.g., drug discovery, aging biomarkers) with 99% accuracy.
* **Use Case:** HRF is intended for high-stakes, first-principles research (AGI, Biology, Physics) where dataset sizes are often limited by experiment cost, but **precision is paramount**.

---
*> "We do not seek to be the fastest. We seek to be the most true." — HRF Research Philosophy*